homeassistant/components/zwave_js/climate.py

```diff
@@ -0,0 +1,332 @@
+    ThermostatMode.FULL_POWER: HVAC_MODE_HEAT,
+}
+
+ZWAVE_SETPOINT_PROPERTY = "setpoint"
+
+
+class OperatingMode(IntEnum):
+    """Enum with all (known/used) Z-Wave OperatingModes."""
+
+    # https://github.com/zwave-js/node-zwave-js/blob/master/packages/zwave-js/src/lib/commandclass/ThermostatOperatingStateCC.ts#L38-L51
+    IDLE = 0
+    HEATING = 1
+    COOLING = 2
+    FAN_ONLY = 3
+    PENDING_HEAT = 4
+    PENDING_COOL = 5
+    VENT_ECONOMIZER = 6
+    AUX_HEATING = 7
+    SECOND_STAGE_HEATING = 8
+    SECOND_STAGE_COOLING = 9
+    SECOND_STAGE_AUX_HEAT = 10
+    THIRD_STAGE_AUX_HEAT = 11
+
+
+HVAC_CURRENT_MAP: Dict[int, str] = {
+    OperatingMode.IDLE: CURRENT_HVAC_IDLE,
+    OperatingMode.PENDING_HEAT: CURRENT_HVAC_IDLE,
+    OperatingMode.HEATING: CURRENT_HVAC_HEAT,
+    OperatingMode.PENDING_COOL: CURRENT_HVAC_IDLE,
+    OperatingMode.COOLING: CURRENT_HVAC_COOL,
+    OperatingMode.FAN_ONLY: CURRENT_HVAC_FAN,
+    OperatingMode.VENT_ECONOMIZER: CURRENT_HVAC_FAN,
+    # not defined in ozw integration but we will provide them anyway
+    OperatingMode.AUX_HEATING: CURRENT_HVAC_HEAT,
+    OperatingMode.SECOND_STAGE_HEATING: CURRENT_HVAC_HEAT,
+    OperatingMode.SECOND_STAGE_COOLING: CURRENT_HVAC_COOL,
+    OperatingMode.SECOND_STAGE_AUX_HEAT: CURRENT_HVAC_HEAT,
+    OperatingMode.THIRD_STAGE_AUX_HEAT: CURRENT_HVAC_HEAT,
+}
+
+
+async def async_setup_entry(
+    hass: HomeAssistant, config_entry: ConfigEntry, async_add_entities: Callable
+) -> None:
+    """Set up Z-Wave climate from config entry."""
+    client: ZwaveClient = hass.data[DOMAIN][config_entry.entry_id][DATA_CLIENT]
+
+    @callback
+    def async_add_climate(info: ZwaveDiscoveryInfo) -> None:
+        """Add Z-Wave Climate."""
+        entities: List[ZWaveBaseEntity] = []
+        entities.append(ZWaveClimate(client, info))
+
+        async_add_entities(entities)
+
+    hass.data[DOMAIN][config_entry.entry_id][DATA_UNSUBSCRIBE].append(
+        async_dispatcher_connect(
+            hass, f"{DOMAIN}_add_{CLIMATE_DOMAIN}", async_add_climate
+        )
+    )
+
+
+class ZWaveClimate(ZWaveBaseEntity, ClimateEntity):
+    """Representation of a Z-Wave climate."""
+
+    def __init__(self, client: ZwaveClient, info: ZwaveDiscoveryInfo) -> None:
+        """Initialize lock."""
+        super().__init__(client, info)
+        self._hvac_modes: Dict[str, Optional[int]] = {}
+        self._hvac_presets: Dict[str, Optional[int]] = {}
+        self.on_value_update()
+
+    @callback
+    def on_value_update(self) -> None:
+        """Call when a watched value is added or updated."""
+        self._current_mode_setpoint_values = self._get_current_mode_setpoint_values()
+        if not self._hvac_modes:
+            self._set_modes_and_presets()
+
+    def _get_current_mode_setpoint_values(self) -> List[ZwaveValue]:
+        """Return a list of current setpoint Z-Wave value(s)."""
+        setpoint_keys: List[str]
+        if not self.info.primary_value:
```

Comment:
I don't think that `info.primary_value` is optional.

Reply:
good point. Is there a better primary value to use? OZW comments indicated that there are cases when the mode is not exposed. I could use a setpoint value instead?

Reply:
I don't know the details of the upstream device behavior here. Someone with better Z-Wave knowledge should comment.

Reply:
fair enough. I read the device class specification ahead of the PRs and I explicitly only supported device classes that have the Thermostat Mode command class, so I think we can stick with it and enhance later if needed